ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/PubsubMessageToTableRow.java

```diff
@@ -58,536 +23,91 @@
-      .setMaxAttempts(12) // Defaults to 6
-      .setTotalTimeout(org.threeten.bp.Duration.ofSeconds(120)) // Defaults to 50 seconds
-      .build();
-
   private final ValueProvider<List<String>> strictSchemaDocTypes;
   private final ValueProvider<String> schemasLocation;
   private final ValueProvider<TableRowFormat> tableRowFormat;
-  private final KeyByBigQueryTableDestination keyByBigQueryTableDestination;
 
   // We'll instantiate these on first use.
-  private transient Cache<TableReference, Schema> tableSchemaCache;
-  private transient Cache<String, String> normalizedNameCache;
-  private transient BigQuerySchemaStore schemaStore;
-  private transient BigQuery bqService;
+  private transient PubsubMessageToObjectNode format;
 
   private PubsubMessageToTableRow(ValueProvider<List<String>> strictSchemaDocTypes,
-      ValueProvider<String> schemasLocation, ValueProvider<TableRowFormat> tableRowFormat,
-      KeyByBigQueryTableDestination keyByBigQueryTableDestination) {
+      ValueProvider<String> schemasLocation, ValueProvider<TableRowFormat> tableRowFormat) {
     this.strictSchemaDocTypes = strictSchemaDocTypes;
     this.schemasLocation = schemasLocation;
     this.tableRowFormat = tableRowFormat;
-    this.keyByBigQueryTableDestination = keyByBigQueryTableDestination;
   }
 
-  /**
-   * Given a KV containing a destination and a message, return the message content as a {@link
-   * TableRow} ready to pass to {@link org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO}.
-   */
-  public TableRow kvToTableRow(KV<TableDestination, PubsubMessage> kv) {
-    PubsubMessage message = kv.getValue();
+  private PubsubMessageToObjectNode createFormat() {
     switch (tableRowFormat.get()) {
       case raw:
-        return rawTableRow(message);
+        return PubsubMessageToObjectNode.Raw.of();
       case decoded:
-        return decodedTableRow(message);
+        return PubsubMessageToObjectNode.Decoded.of();
       case payload:
       default:
-        String tableSpec = kv.getKey().getTableSpec();
-        TableReference ref = BigQueryHelpers.parseTableSpec(tableSpec);
-        try {
-          return payloadTableRow(message, ref, tableSpec);
-        } catch (IOException e) {
-          throw new UncheckedIOException(e);
-        }
-    }
-  }
-
-  /**
-   * Turn the message into a TableRow that contains (likely gzipped) bytes as a "payload" field,
-   * which is the format suitable for the "raw payload" tables in BigQuery that we use for
-   * errors and recovery from pipeline failures.
-   *
-   * <p>We include all attributes as fields. It is up to the configured schema for the destination
-   * table to determine which of those actually appear as fields; some of the attributes will be
-   * thrown away.
-   */
-  @VisibleForTesting
-  static TableRow rawTableRow(PubsubMessage message) {
-    TableRow tableRow = new TableRow();
-    message.getAttributeMap().forEach(tableRow::set);
-    tableRow.set(PAYLOAD, message.getPayload());
```

Comment:
because this PR converts `ObjectNode` to `TableRow`, the value for `payload` will change from `byte[]` to base64 encoded `String`, and I still need test that this doesn't change the resulting value in BigQuery.

Reply:
confirmed that tests cover this in https://github.com/mozilla/gcp-ingestion/pull/1309/commits/21614b68001c78b2177d4e924668cfe0e15f3393